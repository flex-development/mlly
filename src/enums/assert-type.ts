/**
 * @file Enums - AssertType
 * @module mlly/enums/AssertType
 */

/**
 * `import` assertion types.
 *
 * ::: info
 * It's unclear whether or not the HTML spec will require assertion types for
 * WebAssembly modules.
 * :::
 *
 * @see https://github.com/WebAssembly/esm-integration/issues/42
 *
 * @enum {Lowercase<string>}
 */
enum AssertType {
  IMPLICIT = 'javascript',
  JSON = 'json'
}

export default AssertType
