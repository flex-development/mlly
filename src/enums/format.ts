/**
 * @file Enums - Format
 * @module mlly/enums/Format
 */

/**
 * Module formats.
 *
 * @enum {Lowercase<string>}
 */
enum Format {
  BUILTIN = 'builtin',
  COMMONJS = 'commonjs',
  JSON = 'json',
  MODULE = 'module',
  WASM = 'wasm'
}

export default Format
