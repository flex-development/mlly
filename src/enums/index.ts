/**
 * @file Entry Point - Enums
 * @module mlly/enums
 */

export { default as Format } from './format'
export { default as SpecifierKind } from './kind-specifier'
export { default as StatementKind } from './kind-statement'
export { default as SyntaxKind } from './kind-syntax'
