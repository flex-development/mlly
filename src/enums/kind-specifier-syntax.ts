/**
 * @file Enums - SpecifierSyntaxKind
 * @module mlly/enums/SpecifierSyntaxKind
 */

/**
 * Module specifier syntax kinds.
 *
 * @enum {Lowercase<string>}
 */
enum SpecifierSyntaxKind {
  DYNAMIC = 'dynamic',
  STATIC = 'static'
}

export default SpecifierSyntaxKind
