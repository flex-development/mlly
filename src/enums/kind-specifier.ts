/**
 * @file Enums - SpecifierKind
 * @module mlly/enums/SpecifierKind
 */

/**
 * Specifier kinds.
 *
 * @enum {Lowercase<string>}
 */
enum SpecifierKind {
  DYNAMIC = 'dynamic',
  STATIC = 'static'
}

export default SpecifierKind
