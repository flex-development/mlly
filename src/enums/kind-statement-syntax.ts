/**
 * @file Enums - StatementSyntaxKind
 * @module mlly/enums/StatementSyntaxKind
 */

/**
 * CommonJS and ESM statement syntax kinds.
 *
 * @enum {Lowercase<string>}
 */
enum StatementSyntaxKind {
  DECLARATION = 'declaration',
  DEFAULT = 'default',
  DEFAULT_WITH_NAMED = 'default-with-named',
  DEFAULT_WITH_NAMESPACE = 'default-with-namespace',
  DYNAMIC = 'dynamic',
  LIST = 'list',
  NAMED = 'named',
  NAMESPACE = 'namespace',
  REQUIRE = 'require',
  SIDE_EFFECT = 'side-effect'
}

export default StatementSyntaxKind
