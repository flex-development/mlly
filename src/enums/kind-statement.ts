/**
 * @file Enums - StatementKind
 * @module mlly/enums/StatementKind
 */

/**
 * CommonJS and ESM statement kinds.
 *
 * @enum {Lowercase<string>}
 */
enum StatementKind {
  EXPORT = 'export',
  IMPORT = 'import',
  REQUIRE = 'require'
}

export default StatementKind
