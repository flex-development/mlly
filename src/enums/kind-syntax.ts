/**
 * @file Enums - SyntaxKind
 * @module mlly/enums/SyntaxKind
 */

/**
 * CommonJS and ESM statement syntax kinds.
 *
 * @enum {Lowercase<string>}
 */
enum SyntaxKind {
  DECLARATION = 'declaration',
  DEFAULT = 'default',
  DEFAULT_WITH_NAMED = 'default-with-named',
  DEFAULT_WITH_NAMESPACE = 'default-with-namespace',
  DYNAMIC = 'dynamic',
  LIST = 'list',
  NAMED = 'named',
  NAMESPACE = 'namespace',
  REQUIRE = 'require',
  SIDE_EFFECT = 'side-effect'
}

export default SyntaxKind
