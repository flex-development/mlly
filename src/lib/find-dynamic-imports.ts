/**
 * @file findDynamicImports
 * @module mlly/lib/findDynamicImports
 */

import { SpecifierKind, StatementKind, SyntaxKind } from '#src/enums'
import type { DynamicImport } from '#src/interfaces'
import validateString from '#src/internal/validate-string'
import type { NodeError } from '@flex-development/errnode'
import { DYNAMIC_IMPORT_REGEX } from '@flex-development/import-regex'

/**
 * Finds all dynamic import statements in `code`. Ignores matches in comments.
 *
 * @see {@linkcode DynamicImport}
 * @see https://regex101.com/r/PTPAvU
 * @see https://developer.mozilla.org/docs/Web/JavaScript/Reference/Operators/import
 *
 * @param {string} [code=''] - Code to evaluate
 * @return {DynamicImport[]} Dynamic import statement objects
 * @throws {NodeError<TypeError>} If `code` is not a string
 */
const findDynamicImports = (code: string = ''): DynamicImport[] => {
  validateString(code, 'code')

  return [...code.matchAll(DYNAMIC_IMPORT_REGEX)].map(match => {
    const { 0: code = '', index: start = 0, groups = {} } = match
    const {
      expression = '',
      imports = '',
      options = '',
      specifier = ''
    } = groups

    return {
      code,
      end: start + code.length,
      expression,
      imports:
        imports === ''
          ? []
          : imports
              .replace(/^{|}$/g, '')
              .split(',')
              .map(e => e.trim())
              .filter(e => e.length > 0),
      kind: StatementKind.IMPORT,
      options,
      specifier: specifier.replace(/^["']|["']$/g, ''),
      specifier_kind: /^["']/g.test(specifier)
        ? SpecifierKind.STATIC
        : SpecifierKind.DYNAMIC,
      start,
      syntax: SyntaxKind.DYNAMIC
    }
  })
}

export default findDynamicImports
