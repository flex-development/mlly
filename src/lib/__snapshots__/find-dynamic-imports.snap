// Vitest Snapshot v1

exports[`unit:lib/findDynamicImports > should return DynamicImport object array 1`] = `
[
  {
    "code": "await import('foo')",
    "end": 19,
    "expression": "import('foo')",
    "imports": [],
    "kind": "import",
    "options": "",
    "specifier": "foo",
    "specifier_kind": "static",
    "start": 0,
    "syntax": "dynamic",
  },
  {
    "code": "await import(\\"/modules/my-module.js\\")",
    "end": 57,
    "expression": "import(\\"/modules/my-module.js\\")",
    "imports": [],
    "kind": "import",
    "options": "",
    "specifier": "/modules/my-module.js",
    "specifier_kind": "static",
    "start": 20,
    "syntax": "dynamic",
  },
  {
    "code": "await import('foo.json', { assert: { type: 'json' } })",
    "end": 112,
    "expression": "import('foo.json', { assert: { type: 'json' } })",
    "imports": [],
    "kind": "import",
    "options": "{ assert: { type: 'json' } }",
    "specifier": "foo.json",
    "specifier_kind": "static",
    "start": 58,
    "syntax": "dynamic",
  },
  {
    "code": "const foo = await import(bar)",
    "end": 143,
    "expression": "import(bar)",
    "imports": [
      "foo",
    ],
    "kind": "import",
    "options": "",
    "specifier": "bar",
    "specifier_kind": "dynamic",
    "start": 114,
    "syntax": "dynamic",
  },
  {
    "code": "myModule = await import(\\"module-used-on-server\\")",
    "end": 247,
    "expression": "import(\\"module-used-on-server\\")",
    "imports": [
      "myModule",
    ],
    "kind": "import",
    "options": "",
    "specifier": "module-used-on-server",
    "specifier_kind": "static",
    "start": 199,
    "syntax": "dynamic",
  },
  {
    "code": "myModule = await import(\\"module-used-in-browser\\")",
    "end": 309,
    "expression": "import(\\"module-used-in-browser\\")",
    "imports": [
      "myModule",
    ],
    "kind": "import",
    "options": "",
    "specifier": "module-used-in-browser",
    "specifier_kind": "static",
    "start": 260,
    "syntax": "dynamic",
  },
  {
    "code": "import(\`/modules/module-\${index}.js\`)",
    "end": 407,
    "expression": "import(\`/modules/module-\${index}.js\`)",
    "imports": [],
    "kind": "import",
    "options": "",
    "specifier": "\`/modules/module-\${index}.js\`",
    "specifier_kind": "dynamic",
    "start": 370,
    "syntax": "dynamic",
  },
  {
    "code": "const {
  default: myDefault,
  addFive,
  addFour,
  addThree,
  addTwo,
  squareFive,
  squareFour,
  squareThree,
  squareTwo
} = await import('./lib')",
    "end": 566,
    "expression": "import('./lib')",
    "imports": [
      "default: myDefault",
      "addFive",
      "addFour",
      "addThree",
      "addTwo",
      "squareFive",
      "squareFour",
      "squareThree",
      "squareTwo",
    ],
    "kind": "import",
    "options": "",
    "specifier": "./lib",
    "specifier_kind": "static",
    "start": 412,
    "syntax": "dynamic",
  },
  {
    "code": "import(\\"/modules/my-module.js\\")",
    "end": 794,
    "expression": "import(\\"/modules/my-module.js\\")",
    "imports": [],
    "kind": "import",
    "options": "",
    "specifier": "/modules/my-module.js",
    "specifier_kind": "static",
    "start": 763,
    "syntax": "dynamic",
  },
]
`;
